### PR TITLE
Persistant filesystem service

### PR DESCRIPTION
**Pull request by jfarrell**
Updates the filesystem service to match all other services with a node, worker, and gateway. Also adds the ability to serialize and snapshot a filesystem and rollback.

### CODE REVIEW

File: filesystem/Gemfile [hunk 1]
```diff
@@ -12,14 +12,17 @@ gem "thin"
 
 gem 'vcap_common', :require => ['vcap/common', 'vcap/component'], :git => 'git://github.com/cloudfoundry/vcap-common.git', :ref => '68d86f57fb'
 gem 'vcap_logging', :require => ['vcap/logging'], :git => 'git://github.com/cloudfoundry/common.git', :ref => 'b96ec1192'
+<<<<<<< HEAD
```
Comment (drnic) [inline filesystem/Gemfile]: Haha merged git conflict :P

**Comment by andreacampi** [comment]
I love the intent (adding snapshotting etc) but I have a few questions (I will add them as line notes).
Also, I do wish you had split this into smaller commits--it's quite hard to see what is purely refactoring from actual changes.

File: filesystem/lib/filesystem_service/filesystem_node.rb [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+  def dir_size(service_id)
+    path = get_dir_path(service_id)
+    `du -s #{path}`.to_i
+  rescue => e
+    @logger.error("dir_size(#{service_id}): #{e}")
+    -1
+  end
+
+  def serve_redis
```
Comment (andreacampi) [inline filesystem/lib/filesystem_service/filesystem_node.rb]: What's the purpose of this method? It looks like custom functionality.

You are using eval on values that are coming from Redis, that's potentially insecure.

**Comment by andreacampi** [comment]
Would it make sense to also add backup/restore? Personally I wouldn't use it, I think, but I wonder.

**Comment by jfarrell** [comment]
@drnic 
- Thanks, missed that. will take care of it

@andreacampi 
- Backup, restore and quota coming in another patch. 
- Function dir_size is for statistics which i can pull out for now.

**Comment by jfarrell** [comment]
Updated pull request based on comments

**Comment by kushmerick** [comment]
i like the new functionality of asasigning instances to nodes that use their local filesystem.  But this is definitely not suitable for all CF installations.  How come you trashed the extensible/pluggable backend functionality and specifically the NFS and local (gateway filesystem) options?   Can you layer your new option on top of the existing options rather than replace them?

**Comment by jfarrell** [comment]
NFS and local options are still available, they have been moved into the filesystem_driver, no need to subclass for only 3 functions which share so much common code within each function. This change will make it easier to add new filesystem types as necessary within the driver

**Comment by dsboulder** [comment]
@jfarrell we are actually still using Gerrit reviews for vcap-services, not github pull requests.  I'm not going to close this though, because I suspect we'll be off Gerrit this week and can re-evaluate your request then.

**Comment by cf-services** [comment]
@jfarrell I've noticed that you don't have tests for most of the new ruby code that you introduced.  We require at least unit tests for code changes.  In addition, we are probably moving filesystem service to a community supported "contrib" repository, so if you'd like this accepted please add tests and resubmit to the contrib repo which should appear this week.

**Pull request closed without merging**